[브루니] 로그인 / 게시글 권한 부여

23Yong opened the pull request:
안녕하세요! 남은 3주간 잘 부탁드립니다. 🙇‍♂️

## 주요 작업
- [x] 로그인 & 로그아웃 (세션사용)
- [x] 유저 수정시 로그인된 사용자와 일치하는지 확인 후 수정
- [x] 권한이 있는 사용자만 아래 내용을 작업할 수 있도록 구현
  - [x] 게시글 작성
  - [x] 게시글 상세조회
  - [x] 게시글 수정
  - [x] 게시글 삭제
- [x] 컨트롤러(ArticleController, UserController) 테스트코드 작성
- [x] 배포 : [배포url](http://ec2-43-201-18-171.ap-northeast-2.compute.amazonaws.com:8080/)

## 학습 키워드
- 인증과 인가
- 쿠키와 세션
- 인터셉터

## 더 학습해야할 사항
- `XXXMemoryRepository` 클래스들은 테스트코드에서만 사용되고 있는데 계속 관리해 주는 것이 맞는지 모르겠습니다.
  - 이 클래스들을 test 패키지로 이동시키는 것이 더 적합해보입니다. 
- 세션은 서버단에서 관리가 되는 것을 학습했습니다. 하지만 따로 DB 테이블을 설계하고 저장하지 않는데도 불구하고 어디에 저장되는지 찾아보고 있습니다. 톰캣 내부에서 관리가 된다고 하는데 정확히 이해가 되지 않아 좀 더 학습해보겠습니다.
- 서버가 여러 개가 되었을 때 어떻게 세션을 관리해야할까?
  - 관련해서 `세션 클러스터링` 키워드가 나왔는데 학습해봐야할 것 같습니다.

## 고민사항
- 게시글을 생성/읽기/수정/삭제를 수행할 때 세션을 확인하기 위해 추가적인 I/O 연산이 필요한데 이는 세션을 사용함에 있어서 어쩔 수 없는 것인지, 더 나은 방법이 있는지 고민이 됩니다. 이를 개선하기 위해 세션 말고 토큰방식을 많이 사용한다는데 현재는 어쩔 수 없는 부분일까요?

23Yong committed:
docs : session 설정을 위한 설정 변수 추가

23Yong committed:
docs : 구현 계획 작성

23Yong committed:
feat : 로그인 기능 구현

`HttpSession`을 이용해 로그인 기능을 구현
로그인 시 "로그아웃", "개인정보수정"이 표시되도록 수정
로그인 상태가 아니라면 "로그인", "회원가입"이 표시되도록 수정

23Yong committed:
feat : 자신의 개인정보만 수정 가능하도록 구현

로그인한 상태의 사용자는 자신의 정보를 수정할 수 있도록 수정

23Yong committed:
docs : 구현 계획 작성

23Yong committed:
feat : 로그아웃 구현

23Yong committed:
feat : 권한이 없는 사용자는 게시글 작성, 상세내용 조회를 못하도록 기능 구현

인터셉터를 활용해 session이 존재하지 않을 경우 로그인 창으로 리다이렉트

23Yong committed:
refactor : `ArticleDto` 인스턴스 생성 시마다 `DateTimeFormatter` 인스턴스가 생성되지 않도록 수정

23Yong committed:
feat : 글 작성화면에서 글쓴이를 제거하고 게시글을 작성할 수 있도록 기능 구현

23Yong committed:
refactor : `ArgumentResolver`를 활용해서 session에서 userId를 추출

23Yong committed:
feat : 게시글 수정 폼 구현

- 자신의 게시글이 아닌 경우 접근할 수 없도록 구현

23Yong committed:
feat : 게시글 수정 기능 구현

23Yong committed:
feat : 게시글 삭제 기능 구현

23Yong committed:
refactor : `@Login` 애노테이션으로 세션에서 `userId`를 가져오도록 수정

23Yong committed:
chore : url 수정 /questions -> /articles

23Yong committed:
chore : 메서드명 수정

- 게시글을 수정하는 것인데 editProfile 이라 되어 있어서 editArticle 로 수정

23Yong committed:
feat : 세션의 `userId`와 수정하려는 회원의 `userId`가 일치하는지 검증

세션의 `userId`와 수정하려는 회원의 `userId`가 일치할 경우에만 수정하도록 기능 추가

23Yong committed:
refactor : 세션이 `NULL`이 아니도록 `NULL` 검증코드 추가

23Yong committed:
test : ArticleController 테스트코드 작성

23Yong committed:
test : UserController 테스트코드 작성

23Yong committed:
fix : 시퀀스 번호를 static 필드에서 인스턴스 필드로 수정

스레드끼리 공유되지 않아야 하는데 static 인 것을 인스턴스 필드로 수정

23Yong committed:
feat : 게시글이 존재하지 않으면 예외를 던지도록 기능 구현

23Yong committed:
test : ArticleService 테스트코드 작성

janeljs commented on src/main/java/kr/codesqaud/cafe/common/resolver/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,36 @@
+	@Override
+	public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer,
+		NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
+		HttpServletRequest request = (HttpServletRequest)webRequest.getNativeRequest();
```
항상 인텔리제이 코드정렬하는 습관을 들이시면 좋은데요~ 
Actons on save에서 저장하면 수행할 액션을 지정해보세요. 

<img width="444" alt="image" src="https://user-images.githubusercontent.com/68000537/232172023-cb4f1f07-9582-4b2b-b957-624016381759.png">

+ 저는 `optimize imports on the fly`도 지정해두었답니다ㅎㅎ

23Yong replied:
넵! 적용해서 수정하겠습니다!
`네이버 캠퍼스 핵데이 Java 코딩 컨벤션`을 적용하겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/common/resolver/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,36 @@
+			HttpSession session = Objects.requireNonNull(request.getSession(false));
+			return session.getAttribute("sessionedUser");
+		} catch (final NullPointerException e) {
+			throw new InvalidSessionException(e);
```
Throwable에 NullPointerException을 원인 예외로 던져주셨군요 👍🏻

janeljs commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -19,15 +28,37 @@ public ArticleController(ArticleService articleService) {
-	public String posting(@ModelAttribute final PostingRequest request) {
-		articleService.posting(request);
+	@PostMapping
+	public String posting(@ModelAttribute final PostingRequest request, @Login final String userId) {
```
메서드 이름은 보통 동사로 적어준답니다~ㅎㅎ

23Yong replied:
넵! `posting` 에서 `post`로 수정하도록 하겠습니다~!

janeljs commented on src/main/java/kr/codesqaud/cafe/controller/dto/ArticleDto.java:
```diff
@@ -7,6 +7,8 @@
 
 public class ArticleDto {
```
패키지만으로도 이미 dto 클래스임이 잘 드러나기 때문에 dto를 꼭 클래스 이름에 명시해주실 필요는 없답니다~ㅎㅎ 
저는 상황에 맞게 `request` `reponse` `param` 등을 붙여주고 있어요ㅎㅎ

23Yong replied:
넵! 단순 파라미터 전달용이기 때문에 `ArticleParam`으로 수정하겠습니다~!

janeljs replied:
nit: 파라미터라기 보다는 article이 생성되는 것에 대한 요쳥 body를 전달하는 부분이기 때문에 ArticleRequest나 ArticleCreationRequest도 괜찮을 것 같아요~
param은 get 요청 시 검색이나 정렬을 위한 파라미터를 담고있는 dto에 자주 사용합니다 😄

23Yong replied:
감사합니다! 의미를 잘 알고 사용해야겠네요🥲 수정하겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/common/resolver/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,36 @@
+
+import kr.codesqaud.cafe.exception.InvalidSessionException;
+
+public class LoginUserArgumentResolver implements HandlerMethodArgumentResolver {
```
`@RequestAttribute`를 이용하면 따로 argument resolver를 등록해주지 않아도 request attribute에 접근할 수 있답니다~ㅎㅎ

23Yong replied:
감사합니다! `@SessionAttribute`를 통해 세션의 attribute에서 값을 바로 가져오도록 해보겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/exception/GlobalExceptionHandler.java:
```diff
@@ -25,6 +25,24 @@ public String handleInvalidPassword(final InvalidPasswordException e, final Mode
 		return "user/edit_form";
 	}
 
+	@ExceptionHandler(InvalidLoginInfoException.class)
+	public String handleInvalidLoginInfo(final InvalidLoginInfoException e, final Model model) {
+		model.addAttribute("error", e.getMessage());
+		return "user/login";
+	}
+
+	@ExceptionHandler(InvalidSessionException.class)
+	public String handleInvalidSession(final InvalidSessionException e, final Model model) {
+		model.addAttribute("error", e.getMessage());
+		return "error";
+	}
+
+	@ExceptionHandler(NoAuthorizationException.class)
+	public String handleNoAuthorization(final NoAuthorizationException e, final Model model) {
+		model.addAttribute("error", e.getMessage());
+		return "error";
+	}
```
셋 다 내부적으로 동일한 처리를 하고 있는 것 같은데요~ `@ExceptionHandler` 내부에 배열로 세 가지 예외를 모두 넘기거나, 적절한 단위로 상위 예외를 정의하여 상속하는 구조로 구현하고 상위 예외만 잡아줘도 괜찮을 것 같습니다ㅎㅎ
```java
@ExceptionHandler({InvalidSessionException.class, NoAuthorizationException.class, InvalidLoginInfoException.class})
```

23Yong replied:
넵! `@ExceptionHandler` 내부에 배열로 세 가지 예외를 모두 넘기는 방식으로 수정하겠습니다! 감사합니다ㅎㅎ

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/ArticleJdbcRepository.java:
```diff
@@ -52,4 +52,15 @@ public Optional<Article> findById(final Long id) {
+
+	@Override
+	public void update(final Article article) {
+		jdbcTemplate.update("UPDATE article SET title = ?, content = ? WHERE id = ?",
```
`NamedParameterJdbcTemplate`을 이용하는 방식으로 개선해볼 수도 있을 것 같습니다~ㅎㅎ

23Yong replied:
넵! `NamedParameterJdbcTemplate`을 사용함으로써 sql 인자 위치에 대한 순서와 가독성을 떨어뜨리는 문제를 해결해 보겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/ArticleMemoryRepository.java:
```diff
@@ -33,4 +33,14 @@ public List<Article> findAll() {
 		return Optional.ofNullable(articleRepository.get(id));
 	}
+
+	@Override
+	public void update(final Article article) {
+		articleRepository.put(article.getId(), article);
+	}
+
+	@Override
+	public void deleteById(final Long id) {
+		articleRepository.remove(id);
+	}
```
> XXXMemoryRepository 클래스들은 테스트코드에서만 사용되고 있는데 계속 관리해 주는 것이 맞는지 모르겠습니다

더 이상 사용하지 않고 있다면 삭제하셔도 됩니다~ㅎㅎ

23Yong replied:
넵! 감사합니답!ㅎㅎ

janeljs commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -39,4 +40,26 @@ public ArticleDto findById(final Long id) {
+			.orElseThrow(NoAuthorizationException::new);
+	}
+
+	@Transactional
```
`@Transactional`을 사용해주셨네요~ㅎㅎ `@Transactional`을 사용하면 어떤 효과가 있을까요?ㅎㅎ

23Yong replied:
선언적으로 트랜잭션을 지정할 수 있게 되면서 비즈니스 로직에서 트랜잭션 처리 코드들이 모두 제거될 수 있게 되었습니다.
`@Transactional` 애노테이션이 달려있는 메서드 실행 중 예외(기본으로는 UnchekcedException)가 발생시 모두 롤백되게 되고 그렇지 않다면 모두 성공하게 됩니다!
그런데 왜 스프링은 `CheckedException`의 경우는 롤백되지 않게 설정했는지 모르겠습니다. `UncheckedException`의 경우는 예상치 못한 예외이고 `ChekcedException`은 예상된 예외이기 때문에 그런 걸까요?

janeljs replied:
CheckedException은 이미 언어레벨에서 try~catch 사용을 강제하기 때문에 예외 처리를 이미 하고있고, UncheckedException은 예외 처리의 책임을 강제하지 않기 때문이라고 생각합니다ㅎㅎ

janeljs commented on src/test/java/kr/codesqaud/cafe/controller/ArticleControllerTest.java:
```diff
@@ -0,0 +1,244 @@
+	void givenPostingData_whenPosting_thenRedirectsHome() throws Exception {
+		// given
+		MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
+		body.add("title", "테스트코드는 어려워!");
+		body.add("content", "왜 어려운거지?");
```
😄

janeljs commented on src/test/java/kr/codesqaud/cafe/controller/ArticleControllerTest.java:
```diff
@@ -0,0 +1,244 @@
+	@Test
+	void givenNotEqualUserId_whenShowArticleEditPage_thenReturnsErrorView() throws Exception {
+		// given
+		doThrow(NoAuthorizationException.class).when(articleService).validateHasAuthorization(anyLong(), anyString());
```
정렬이 깨진 것 같아요~ㅎㅎ

23Yong replied:
엇 수정하겠습니다~! 감사합니다~ㅎㅎ

janeljs commented on src/test/java/kr/codesqaud/cafe/service/ArticleServiceTest.java:
```diff
@@ -81,4 +88,88 @@ void givenNotExistsArticleId_whenFindById_thenThrowsException() {
+	}
+
+	@DisplayName("게시글을 수정할 때")
+	@Nested
```
오~ `@Nested`를 활용해서 테스트를 적절히 분리해주셨군오 💯

janeljs commented on src/test/java/kr/codesqaud/cafe/service/ArticleServiceTest.java:
```diff
@@ -81,4 +88,88 @@ void givenNotExistsArticleId_whenFindById_thenThrowsException() {
+
+			// then
+			Article article = articleService.findById(1L).toEntity();
+			SoftAssertions.assertSoftly(softAssertions -> {
```
soft assertions 사용해주셨군요~ㅎㅎ assertSoftly를 사용할 경우 assertAll을 사용하는 것과는 어떤 차이가 있나요?ㅎㅎ

23Yong replied:
jupiter의 `assertAll`는 다음과 같이 실행됩니다.
- `assertAll` 내부의 단언문들이 실행되던 도중 발생한 예외는 모두 그룹화되어 `MultipleFailuresError`에 담아 한 번 보고됩니다. 이때 테스트 도중 하나가 실패하더라도 나머지 테스트들은 계속 수행되게 됩니다.
- 모든 그룹화된 예외들은 `MultipleFailuresError`에 담겨지게 됩니다.

assertj의 `softAssertly`는 다음과 같이 실행됩니다.
- `SoftAssertions`는 AssertJ 객체의 프록시 객체를 제공해서 어설션 실패를 캐치하고 저장합니다. jupiter의 `assertAll`과 마찬가지로 테스트 도중 하나가 실패하더라도 나머지 테스트들은 계속 수행되게 됩니다.
- 이후 `SoftAssertionsProvider.assertAll()` 메서드로 캐치한 예외들을 처리하게 됩니다.

둘을 실행시킨 결과 
`asserAll`은 `MultipleFailuresError`를 던졌고 `softAssertly`는 `AssertJMultipleFailuresError`를 던지는 것을 확인했습니다.
![image](https://user-images.githubusercontent.com/66981851/232271560-31801381-ca56-4fe2-947d-319e6d7a83c7.png)
![image](https://user-images.githubusercontent.com/66981851/232271686-a8792cd4-230f-49b0-be5d-116f44b71a72.png)

jupiter의 `assertAll`은 내부 단언문들을 하나씩 실행시키고, assertj의 `softAssertly`는 AssertJ 객체를 프록시로 래핑해 예외를 잡고, 이를 assertAll() 메서드에서 한꺼번에 처리하도록 하는 것에서 차이가 있는 것 같습니다.. 어떻게 프록시로 래핑하는지는  추후에 좀 더 학습해야할 것 같습니다ㅠ

janeljs replied:
굳굳 깊게 학습하셨네요ㅎㅎ 💯

janeljs commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -19,15 +28,37 @@ public ArticleController(ArticleService articleService) {
+	@PostMapping
+	public String posting(@ModelAttribute final PostingRequest request, @Login final String userId) {
+		articleService.posting(
+			new ArticleDto(null, userId, request.getTitle(), request.getContents(), LocalDateTime.now()));
```
현재 dto class가 PostingRequest도 있고, ArticleDto도 있는데요ㅎㅎ post body로 온 request 객체를 service에 그대로 넘기는 방법으로 개선할 수는 없을까요?ㅎㅎ
현재 id도 null이고 `LocalDateTime.now()`도 컨트롤러에서 생성해서 파라미터로 넘길 필요가 없어보여요~

23Yong replied:
넵! 개선해보겠습니다!

janeljs commented on src/test/java/kr/codesqaud/cafe/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,176 @@
+import kr.codesqaud.cafe.exception.NoAuthorizationException;
+import kr.codesqaud.cafe.service.UserService;
+
+@WebMvcTest(UserController.class)
```
UserController.class를 명시해주셨군요~👍🏻 이렇게 명시해주면 어떤 효과가 있을까요?ㅎㅎ

23Yong replied:
어떤 `@Controller` 빈을 application context에 등록해 줄지 명시하는 효과가 있습니다!
만약 `@WebMvcTest`를 사용할 경우 모든 `@Controller` 빈들이 application context에 등록됩니다!

janeljs reviewed (approved):
브루니 안녕하세요~ㅎㅎ 리뷰어 Jane입니다.
전반적으로 코드가 깔끔하게 잘 작성된 것 같아요 💯 
몇 가지 코멘트 남겨드렸는데 학습해보시고 모르는 내용 있으시면 편하게 질문주세요 😄 

> 게시글을 생성/읽기/수정/삭제를 수행할 때 세션을 확인하기 위해 추가적인 I/O 연산이 필요한데 이는 세션을 사용함에 있어서 어쩔 수 없는 것인지, 더 나은 방법이 있는지 고민이 됩니다. 이를 개선하기 위해 세션 말고 토큰방식을 많이 사용한다는데 현재는 어쩔 수 없는 부분일까요?

세션을 확인하기 위해 추가적인 I/O 연산이 필요하다는게 서버의 HttpSession 스토리지에 대한 I/O 비용 말씀하시는 거 맞나요? 맞다면 I/O 비용에 대해 크게 신경 필요는 없을 것 같습니다. 세션을 이용함으로써 로컬 서버의 메모리가 부족해지거나 I/O 연산이 서버의 부하를 초래할 정도면 외부에 별도 세션 스토리지를 구축하는게 맞는 방향일 것 같고.. token 방식을 이용하는게 I/O 연산을 줄이기 위함은 아닌 것으로 알고 있습니다ㅎㅎ (혹시 찾아보신 자료 출처를 공유받을 수 있을까요? 😄 )

23Yong committed:
refactor : [리뷰사항반영] 인텔리제이 코드 정렬 적용

23Yong committed:
refactor : [리뷰사항반영] 메서드명 수정 posting -> post

메서드명은 동사로 나타내는 것이 좋기 때문에 `posting`에서 `post`로 수정

23Yong committed:
refactor : [리뷰사항반영] 클래스명 수정 XXXDto -> XXXParam

패키지에서 이미 `dto`임이 충분히 드러나기 때문에 이를 수정

23Yong committed:
refactor : [리뷰사항반영] `@ExceptionHandler`로 두 가지 예외를 잡도록 수정

`InvalidSessionException`, `NoAuthorizationException` 두 가지 예외 발생 시 모두 동일한 로직을 수행하고 있기 때문에 한 번에 두 가지 예외를 잡도록 수정

23Yong committed:
refactor : [리뷰사항반영] `@SessionAttribute` 애노테이션 사용

ArgumentResolver를 사용하지 않고 `@SessionAttribute`를 사용하도록 수정

23Yong committed:
refactor : [리뷰사항반영] `ArticleParam`을 `PostingRequest`를 사용하도록 수정

- `ArticleController` 클래스에서 `post`메서드 호출 시 바로 `PostingRequest`를 넘기도록 수정

23Yong committed:
refactor : [리뷰사항반영] `JdbcTemplate`을 `NamedParameterJdbcTemplate`으로 수정

- 기존 SQL 인자에 대한 순서가 강제되는 `JdbcTemplate`의 문제를 `NamedParameterJdbcTempalte`을 사용함으로써 해결

23Yong commented:
리뷰주신 사항들을 확인하고 수정했습니다! 
아무생각 없이 사용하고 있던 것들에 대해 다시 한 번 학습해보는 시간을 가질 수 있었습니다. 감사합니다🙇‍♂️
이후 리뷰사항 반영은 최대한 빨리 반영해서 답변 남기도록 해보겠습니다.

---
인증/인가에 있어서 세션이 토큰보다 추가적인 I/O 를 발생시킨다고 생각한 이유는 다음과 같습니다.
- 클라이언트는 세션아이디를 쿠키에 담아 서버에 전송하게 되면 서버는 세션아이디를 세션 저장소에서 가져오기 때문에 I/O연산이 발생하게 된다고 생각했습니다.
- 반면 토큰방식은 토큰 자체가 이미 인증된 문자열이기 때문에 서버에서 이를 디코딩하기만 하면 따로 저장소를 거치지 않기 때문에 추가적인 I/O연산이 발생하지 않는다고 생각했습니다!

https://www.geeksforgeeks.org/session-vs-token-based-authentication/

janeljs commented on src/main/java/kr/codesqaud/cafe/common/interceptor/LoginInterceptor.java:
```diff
@@ -0,0 +1,23 @@
+
+import org.springframework.web.servlet.HandlerInterceptor;
+
+public class LoginInterceptor implements HandlerInterceptor {
```
시간되실 때 Filter와 Interceptor의 차이에 대해서도 한 번 생각해보세요~ㅎㅎ

23Yong replied:
Filter
- Spring Context 영역 밖에서 동작합니다.
- Servlet으로 가기 전에 요청을 가로채 로직을 수행할 수 있습니다.
- 필터를 체이닝할 수 있습니다.
- `DelegatingFilterProxy`를 통해 필터를 빈으로 등록할 수 있습니다.

Interceptor
- Spring Context 영역 내에서 동작합니다.
- `DispatcherServlet`과 `Controller` 사이에서 동작합니다.
- 인터셉터 체이닝할 수 있습니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/config/MvcConfig.java:
```diff
@@ -2,9 +2,12 @@
 
+import kr.codesqaud.cafe.common.interceptor.LoginInterceptor;
+
 @Configuration(proxyBeanMethods = false)
```
`proxyBeanMethods = false`를 선언하면 어떤 효과가 있을까요?ㅎㅎ 추가로 JDK Dynamic Proxy와 CGLib Proxy에 대해서도 학습해보세요~

23Yong replied:
- `proxyBeanMethods = false`를 사용하면 `@Configuration` 애노테이션이 달린 클래스의 프록시 클래스가 생성되지 않는 것으로 알고있습니다. 
- 원래 `@Configuration` 클래스는 내부 `@Bean` 메서드를 호출해 빈을 등록할 때 싱글턴 스콥으로 등록하기 위해 프록시 클래스를 생성해 이를 가능하게 합니다. 하지만 내부 `@Bean` 메서드를 두 번 이상 호출하는 것이 아니면 `proxyBeanMethods = false`를 사용해 프록시 클래스가 생성되지 않도록 하는 것이 자연스럽습니다!

```java
@Configuration
public class Config {
    @Bean
    public A a() {
        return new A();
    }

    @Bean
    public B b() {
        return new B(a());
    }
}
```
위와 같이 사용되는 경우, `a()`를 호출할 때 `A` 빈이 한 번 생성되고 `B`빈을 생성하는 시점에서 `a()`가 한 번 더 호출되어 `A` 빈이 두 번 생성되게 됩니다! 위와 같이 내부에서 `@Bean` 메서드를 호출하는 것이 아니면 `proxyBeanMethods = false`를 사용해 불필요한 프록시 클래스 생성을 막고 코드 자체로 이해할 수 있게 하는 것이 좋다고 알고 있습니다!

---
- `JDK Dynamic Proxy`
  - 인터페이스 기반의 프록시 클래스 생성방법입니다. 런타임 중에 `Reflection`을 사용하여 타깃 클래스의 프록시 클래스를 생성하게 됩니다.
  - `JDK Dynamic Proxy`는 인터페이스를 기반으로 프록시 클래스를 생성하기 때문에 타깃 클래스가 인터페이스를 구현하고 있어야 한다는 점을 주의해야 합니다.
- `CGLib`
  - 상속을 기반으로 한 프록시 클래스 생성방법입니다. `JDK Dynamic Proxy`와는 다르게 런타임 중에 바이트코드를 직접 조작해 프록시 클래스를 생성하게 됩니다.
  - 스프링 부트는 기본적으로 `CGLib` 방식을 통해 스프링 AOP를 지원합니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/config/MvcConfig.java:
```diff
@@ -15,4 +18,13 @@ public void addViewControllers(ViewControllerRegistry registry) {
+		String[] authorizePatterns = new String[] {"/articles/**", "/question/**"};
+
+		registry
+			.addInterceptor(new LoginInterceptor())
```
이렇게 생성해서 넣어주셔도 되지만 의존성 주입받아서 관리할 수도 있답니다~ㅎㅎ

23Yong replied:
넵! `LoginInterceptor`를 빈으로 등록하고 의존성을 주입받는 방식으로 개선하겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -19,15 +27,39 @@ public ArticleController(ArticleService articleService) {
-		articleService.posting(request);
+	@PostMapping
+	public String post(@ModelAttribute final PostingRequest request,
+					   @SessionAttribute(SESSION_USER) final String userId) {
```
👍🏻

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/ArticleJdbcRepository.java:
```diff
@@ -14,27 +19,25 @@
-		this.jdbcTemplate = jdbcTemplate;
+	public ArticleJdbcRepository(DataSource dataSource) {
+		this.jdbcTemplate = new NamedParameterJdbcTemplate(dataSource);
+		this.jdbcInsert = new SimpleJdbcInsert(dataSource)
```
👍🏻

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/ArticleJdbcRepository.java:
```diff
@@ -47,9 +50,22 @@ public List<Article> findAll() {
+		Map<String, Object> params = Map.of("title", article.getTitle(),
+											"content", article.getContent(),
+											"id", article.getId());
+		jdbcTemplate.update("UPDATE article SET title=:title, content=:content WHERE id=:id", params);
```
수정하시라는 건 아니고 upsert라는 개념에 대해서도 공부해보세요~ㅎㅎ

23Yong replied:
넵! upsert에 대해 학습해보겠습니다~ㅎㅎ

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/UserJdbcRepository.java:
```diff
@@ -49,23 +52,31 @@ public List<User> findAll() {
 		try {
 			return Optional.ofNullable(
-				jdbcTemplate.queryForObject("SELECT * FROM user_account WHERE user_id = ?", userMapper, userId));
+				jdbcTemplate.queryForObject("SELECT * FROM user_account WHERE user_id = :userId",
```
쿼리에 `*`를 쓰는 건 좋지 않아요~ㅎㅎ

23Yong replied:
넵! 불필요한 데이터 검색과 모든 컬럼을 가져와 네트워크 사용량이 늘어나고 디스크에서 모든 컬럼을 읽어과야 해 쿼리 실행 속도가 느려져 좋지 않습니다. 명시적으로 각 컬럼의 이름을 지정해 가져오도록 수정하겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/UserJdbcRepository.java:
```diff
@@ -49,23 +52,31 @@ public List<User> findAll() {
+				"SELECT EXISTS (SELECT user_id FROM user_account WHERE user_id = :userId)",
+				Map.of("userId", userId), Boolean.class));
+		} catch (final EmptyResultDataAccessException e) {
+			return false;
```
여기서 false를 return하기 보다는 바로 optional.empty()를 반환하는 건 어떨까요?
```java
if (isExistUserByUserId(user.getUserId())) {
			return Optional.empty();
		}
```

23Yong replied:
넵! 아래와 같이 수정하겠습니다.
```java
private Optional<Boolean> isExistUserByUserId(final String userId) {
		try {
			return Optional.of(Boolean.TRUE.equals(jdbcTemplate.queryForObject(
				"SELECT EXISTS (SELECT user_id FROM user_account WHERE user_id = :userId)",
				Map.of("userId", userId), Boolean.class)));
		} catch (final EmptyResultDataAccessException e) {
			return Optional.empty();
		}
	}
```
`Optional`은 null을 안전하게 처리하기 위한 방법으로 알고 있는데 원시타입인 `false`를 바로 반환하는 것은 어떤 이유에서 일까요? 저는 값이 있다, 없다가 아닌 예외상황을 나타내주기 위해 `Optional.empty()`를 반환하는 것이라고 이해했습니다!

janeljs replied:
헛 제가 왜 이렇게 리뷰를 남겼을까요ㅠㅠ (도저히 원인을 모르겠네요..) 처음에 구현해주신 상태가 훨씬 나은 것 같습니다. 죄송합니다.
boolean 값을 굳이 optional로 return할 필요도 전혀 없고, 메서드 시그니처도 너무 어색하네요.. 롤백해주시면 될 것 같습니다..^^

janeljs commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -23,20 +26,43 @@ public ArticleService(ArticleRepository articleRepository) {
+
+	public void validateHasAuthorization(final Long articleId, final String userId) {
+		articleRepository.findById(articleId)
+			.filter(article -> article.getWriter().equals(userId))
```
`article.getWriter().equals(userId)` 이 로직은 article 내부에 위치해도 좋을 것 같아요~ 왜 그런지 생각해보세요~ㅎㅎ

23Yong replied:
말씀대로 도메인 로직으로 분리하는 것이 좋아보입니다!
서비스 계층에서 도메인 계층으로 `userId`(작성자)를 전달하고 도메인이 같은 작성자인지 확인한 후 값을 반환하도록 하는 것이 관심사를 분리할 수 있는 방법 같습니다.
게시글의 작성자와 전달받은 `userId`가 같은지 확인하는 로직은 게시글 도메인의 역할이기 때문인 것 같습니다.

janeljs replied:
https://medium.com/vattenfall-tech/the-law-of-demeter-by-example-fd7adbf0c324

위에 글도 한 번 읽어보세요~ㅎㅎ

janeljs commented on src/test/java/kr/codesqaud/cafe/service/ArticleServiceTest.java:
```diff
@@ -81,4 +88,88 @@ void givenNotExistsArticleId_whenFindById_thenThrowsException() {
+			articleService.editArticle(1L, request);
+
+			// then
+			Article article = articleService.findById(1L).toEntity();
```
굳이 dto로 리턴한 걸 엔티티로 바꾸셔서 검증하시는 이유가 있을까요?

23Yong replied:
말씀대로 굳이 dto를 엔티티로 검증하는 이유는 없네요..🥲 수정하겠습니다! 감사합니다!🙇‍♂️

janeljs commented on src/main/java/kr/codesqaud/cafe/controller/dto/ArticleParam.java:
```diff
@@ -0,0 +1,40 @@
+		this.createdAt = DATE_TIME_FORMATTER.format(createdAt);
+	}
+
+	public Article toEntity() {
+		return new Article(
+			id,
+			writer,
+			title,
+			content,
+			LocalDateTime.parse(createdAt, DATE_TIME_FORMATTER)
+		);
+	}
```
현재 테스트에서만 사용되고 있는 메서드인 것 같습니다.

23Yong replied:
꼼꼼히 봐주셔서 감사합니다!🥺 제거하겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/common/consts/SessionConst.java:
```diff
@@ -0,0 +1,9 @@
+package kr.codesqaud.cafe.common.consts;
+
+public final class SessionConst {
```
nit: SESSION_USER 상수 관리만을 위해서 유틸 클래스를 만드는 건 살짝 과한 것 같기도 하네요ㅎㅎ `SESSION_USER`가 여러 클래스를 걸쳐 사용되나요?

23Yong replied:
`ArticleController`와 `UserController` 클래스에 사용되고 있습니다!
단순히 하나의 상수 관리를 위한 클래스는 과하다고 생각되네요.. 각각의 클래스에서 상수로 만드는 것이 더 좋은 방법일까요?

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/UserJdbcRepository.java:
```diff
@@ -49,23 +52,31 @@ public List<User> findAll() {
 	public void update(final User user) {
-		jdbcTemplate.update("UPDATE user_account SET password = ?, name = ?, email = ? WHERE user_id = ?",
-			user.getPassword(), user.getName(), user.getEmail(), user.getUserId());
+		Map<String, Object> params = Map.of("password", user.getPassword(),
+											"name", user.getName(),
+											"email", user.getEmail(),
+											"userId", user.getUserId());
```
여기도 뭔가 정렬이 깨진 것 같습니다ㅎㅎ

janeljs commented on src/main/java/kr/codesqaud/cafe/repository/impl/UserJdbcRepository.java:
```diff
@@ -49,23 +52,31 @@ public List<User> findAll() {
+											"email", user.getEmail(),
+											"userId", user.getUserId());
+		jdbcTemplate.update(
+			"UPDATE user_account SET password = :password, name = :name, email = :email WHERE user_id = :useId",
```
useId는 오타일까요?

23Yong replied:
엇.. 감사합니다!🙇‍♂️ 확인해서 수정하겠습니다!

janeljs commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -23,20 +26,43 @@ public ArticleService(ArticleRepository articleRepository) {
-		articleRepository.save(Article.from(request));
+	public void post(final PostingRequest request, final String userId) {
+		articleRepository.save(
+			new Article(null, userId, request.getTitle(), request.getContents(), LocalDateTime.now()));
```
null이랑 `LocalDateTime.now()`는 파라미터로 넘기지는 않아도 되지 않을까요?ㅎㅎ 
userId, title, contents를 받는 static method of를 정의해주거나 PostingRequest를 매개변수로 받는 정적 팩터리 메서드를 정의하는게 좋을 것 같아요~

23Yong replied:
넵! 정적 팩터리 메서드를 정의해서 수정하겠습니다!
```java
public static Article of(final String writer, final String title, final String content) {
	return new Article(writer, title, content);
}
```

janeljs commented:
> 인증/인가에 있어서 세션이 토큰보다 추가적인 I/O 를 발생시킨다고 생각한 이유는 다음과 같습니다.
> - 클라이언트는 세션아이디를 쿠키에 담아 서버에 전송하게 되면 서버는 세션아이디를 세션 저장소에서 가져오기 때문에 I/O연산이 발생하게 된다고 생각했습니다.
> - 반면 토큰방식은 토큰 자체가 이미 인증된 문자열이기 때문에 서버에서 이를 디코딩하기만 하면 따로 저장소를 거치지 않기 때문에 추가적인 I/O연산이 발생하지 않는다고 생각했습니다!

말씀주신대로 세션은 저장소에 할당되는 메모리 & 세션아이디를 저장/조회할 때 I/O 연산이 발생하는 비용이 존재하기는 하는데요.
이 때문에 WAS에 영향이 갈 정도라면 별도로 글로벌 세션 스토리지를 구축하는게 맞을 것 같다는 말씀을 드리고 싶었습니다ㅎㅎ
토큰 방식은 말씀주신대로 jwt에 이미 필요한 정보가 모두 담겨있기 때문에 추가적인 I/O 연산은 발생하지 않지만 연산 비용보다는 무상태성이 더 토큰을 사용하게 되는 이유이지 않을까 싶었어요~!ㅎㅎ 특히 MSA 아키텍처를 취하거나 확장성이 필요한 경우에는 더 유용하겠죠?
